all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+`NumWorkflowTaskPollers` (or routines), what it probably should be is `MaxConcurrentWorkflowTaskPolls`, since the number
+of open RPC calls is actually what's fluctuating.
+
+
+### Slot tuning & Cache Size
+
+Workers have a configurable number of slots for task types. In Core: `max_outstanding_workflow_tasks`,
+`max_outstanding_activities`, and `max_outstanding_local_activities`. These are set by the user and can't change at
+runtime. Autotuning of these values seeks to allow them to change at runtime according to available resources on
+the worker. It is not feasible for us to choose *optimal* values for these at runtime, because we cannot know a-priori
+about how much resources any given workflow or activity task will consume - only the user can (maybe) know this.
+However, we can provide reasonable default autotuning implementations while also allowing users to provide their own.
+
+Additionally, there is a maximum cache size for cached workflows. In Core: `max_cached_workflows`. This is also set by
+users at the moment, and autotuning it is subject to the same lack of knowledge we have as with slot tuning.
+
+## Constraints
+
+There are some invariants that must be maintained:
+
+* The current number of active polls for a task type must be <= the number of free slots for that task type at all
+  times. Otherwise, we can end up receiving a task that won't have anywhere to go until a slot frees. For example,
+  in Core this is expressed by using a permit reservation system.
+* Max outstanding workflow tasks must be <= max cached workflows.
+
+## Proposed interface for slot management
+
+To reiterate, we can't possibly know as well as the user what resources a given workflow or activity task will consume.
+Thus, we need to provide an interface by which users can tell us whether they think it's reasonable to allow the use
+of an additional slot.
+
+The interface definition here is provided in Rust, but need not be copied exactly by other languages. Since users are
+implementing it, we want something idiomatic for them. Rust will likely immediately convert all newly allowed slots into
+a permit struct which will automatically release the slot when dropped (like it does today, without this interface). 
+Languages with destructor semantics should do the same, as permit/slot leaking is easy to do by accident.
+
+There is some mild generic magic going on to represent different types of slots and constrain that the right info type
+is provided. These guarantees can't be compile-time enforced across the language boundary, but are at least useful after
+runtime validation.
+
+```rust
+trait SlotSupplier {
+  type SlotKind: SlotKind;
+  /// Blocks until a slot is available.
+  /// May also return an error if the backing user implementation encounters an error.
+  /// Returned errors will bubble up and cause the worker to shut down.
+  async fn reserve_slot(&self) -> Result<(), anyhow::Error>;
```

Comment:
Sure, that's a good point. It can't (well, shouldn't) be exactly the same type as in the other places though, as it'd mean 90% of the fields on it become optional. Perfectly reasonable to have a new type here though with the fields we know like you mentioned, and then room to add more fields later on.